File: solutions/julia/run-length-encoding/run-length-encoding.jl
using Pipe:@pipe

function encode(s::String)::String
    @pipe eachmatch(r"(.)\1*", s) |>
        Iterators.map(m -> m.match, _) |>
        collect |>
        map(s -> length(s) == 1 ? s[1] : string(length(s), s[1]), _) |>
        string(_...)
end

function decode(s::String)::String
    @pipe eachmatch(r"(\d*)([a-zA-Z\s])", s) |>
        Iterators.map(m -> isempty(m[1]) ? (1, m[2]) : (parse(Int, m[1]), m[2]), _) |>
        collect |>
        map(((n, c),) -> c^n, _) |>
        string(_...)       
end

File: solutions/fsharp/run-length-encoding/RunLengthEncoding.fs
module RunLengthEncoding

open System.Text.RegularExpressions
open System

let encodeOneChar c i =
    if i > 1 then $"{i}{c}"
    else if i = 1 then $"{c}"
    else ""

let encode (input: string) =
    let rec encodeNext (input: char list) (encodeAcc: string) (charAcc: char) (countAcc: int) =
        match input with
        | [] -> $"{encodeAcc}{encodeOneChar charAcc countAcc}"
        | h :: t when h = charAcc -> encodeNext t encodeAcc charAcc (countAcc + 1)
        | h :: t -> encodeNext t $"{encodeAcc}{encodeOneChar charAcc countAcc}" h 1

    encodeNext (Seq.toList input) "" Char.MinValue 0

let expandAbbreviation (abbrev: string) =
    if abbrev.Length = 1 then
        abbrev
    else
        Regex.Matches(abbrev, "(\d+)?(.)").[0].Groups
        |> Seq.toList
        |> List.map (fun m -> m.Value)
        |> (fun s -> List.replicate (int s.[1]) s.[2])
        |> System.String.Concat

let decode input =
    Regex.Matches(input, "(\d+)?[a-zA-Z\s]")
    |> Seq.map ((fun m -> m.Value) >> expandAbbreviation)
    |> System.String.Concat
